port stest to rust

benjaminedwardwebb opened the pull request:
I took the opportunity to learn some rust (something I'd been meaning to do) and port `stest` over from C, as suggested in [this issue][1]. Hopefully the rust looks reasonable, but open to any advice and tips of course.

Since stest was the only remaining C artifact, this pull request also drops all references to the C toolchain from the makefile and build. Yay, less C and simpler build!

[1]: https://github.com/Shizcow/dmenu-rs/issues/45#issuecomment-1360881144

# verification

I wrote some integration tests for `stest`. They're not perfect but they cover a lot of the basic cases.

I also ran modified versions of `dmenu_path` and `dmenu_run` that pointed to the rust-based `stest` built locally on my `x86_64 GNU/Linux` host, and dmenu ran as expected.

benjaminedwardwebb commented on src/stest/tests/set-up-block-special-file:
```diff
@@ -0,0 +1,9 @@
+# main : () -> string
+# Return a path to a block special file that exists on most systems.
+function main {
+	echo "/dev/disk/by-diskseq/1" || return 1
```
I did my best to come up with a file I thought would be an most unix machines, but I'm not the most familiar with linux and don't know how well this works for sure.

Maybe there's an easy way to construct a block special file that we can use instead?

Shizcow replied:
I __believe__ that `/dev/disk/by-diskseq/1` works here. Good job finding this.

benjaminedwardwebb commented on src/stest/tests/set-up-character-special-file:
```diff
@@ -0,0 +1,9 @@
+# main : () -> string
+# Return a path to a character special file that exists on most systems.
+function main {
+	echo "/dev/tty" || return 1
```
Same as above. It bothers me that we're making assumptions about existing files on the file system in this test and not just creating the test conditions ourselves.

Shizcow replied:
Neither `/dev/tty` nor `/dev/disk/by-diskseq/1` are actually files; they are guaranteed to exist by Linux.
> This statement is almost certainly wrong in some crazy edge case, but the assumption is reasonable. It's appropriate to go forward this way and change if someone files a bug down the road.

benjaminedwardwebb replied:
> This statement is almost certainly wrong in some crazy edge case, but the assumption is reasonable. It's appropriate to go forward this way and change if someone files a bug down the road.

I agree. I'll plan to move forward with this until something better comes up.

One crazy edge-case that's on my mind, however, is the [sandboxed environment of a nix build][1]. But that can come later.

[1]: https://discourse.nixos.org/t/survey-how-are-you-using-flakes-today/15494/4

Shizcow replied:
Do nix builds commonly run tests as well? If not, this is a non-issue.

benjaminedwardwebb replied:
I expect that many nix builds inside of the `nixpkgs` repository do end up skipping tests for one reason or another, however it requires an explicit disablement and I think it's probably discouraged.

For example, I disabled the test phase of `dmenu-rs`'s `make` build when checking it into `nixpkgs` [here][1]. There was some discussion in the PR from contributors about ways to emulate X to enable the tests to run, but ultimately it proved too much to hack around. But I hope that can change in the future.

[1]: https://github.com/NixOS/nixpkgs/blob/28319deb5ab05458d9cd5c7d99e1a24ec2e8fc4b/pkgs/applications/misc/dmenu-rs/default.nix#L79

Shizcow commented on src/Cargo.toml:
```diff
@@ -4,4 +4,5 @@ members = [
         "headers",
-]
\ No newline at end of file
+		"stest"
```
Nitpick: Can this have the same indentation?

benjaminedwardwebb replied:
Of course. Sorry about this -- tabs not spaces. :)

On another note, probably we could enforce this stuff with rustfmt or some other pre-commit hook?

Shizcow commented on src/stest/Cargo.toml:
```diff
@@ -0,0 +1,9 @@
+version = "0.0.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Nitpick: Can this comment be removed?

Shizcow commented on src/stest/Cargo.toml:
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "stest"
+version = "0.0.0"
```
When I run the new `stest` binary as `stest --version` it reports `0.0.0`. Can you have `clap` reference `env!("VERSION")` to get the version info when passed `--version`?

Shizcow replied:
You might need to override the `--version` command for this.

benjaminedwardwebb replied:
Extremely pleased to discover that this just works:

```rust
#[command(author, version = env!("VERSION"), about, long_about)]
```

Clap really is awesome :)

Shizcow commented on makefile:
```diff
@@ -40,6 +33,7 @@ man:	config
 	man target/dmenu.1
 
 test:	all
+	cd src && cargo test
```
Thanks for adding tests. This fixes the problem with removing `checkdeps.sh`. Now we can be certain that everything works without needing to test it beforehand.

Shizcow commented on src/stest/tests/set-up-directory:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
```
Is there any reason that all of the files in the `tests` directory are shell scripts, but none of them have an `.sh` extension? Most editors will autodetect this from the shebang, but not all.

benjaminedwardwebb replied:
It's because they're executable.

I feel like it's conventional that an executable file has no extension. I suppose the fact that these are shell scripts can be considered an implementation detail -- they could equally be C binaries and the caller needn't know.

On the other hand, if these scripts were non-executable bash "source code" files that I expected to be loaded with `source` or `.`, I would definitely add a `.bash` extension.

Shizcow commented on src/stest/Cargo.lock:
```diff
@@ -0,0 +1,323 @@
+# This file is automatically @generated by Cargo.
+# It is not intended for manual editing.
+version = 3
```
I have a pending work item to git ignore Cargo.lock, so this is okay to include in the PR.

Shizcow commented on src/stest/src/file.rs:
```diff
@@ -0,0 +1,232 @@
+
+/// Wraps a [PathBuf] for use in stest.
+///
+/// These files are assumed to be on a unix system. Error-handling is simple: always panic.
```
Just unwrapping makes for very messy errors. In any place you *actually* expect to see an error, you should at least `expect("A better error message")`. Realistically, `expect` isn't as good as just printing an error and exiting, as it also prints source line info (not clean).

Shizcow replied:
In the case of this program, I believe the right way of doing it would be to return a `Result<bool, String>` from all of the test functions. `true` or `false` are the test results. `Error` would then be a problem performing the test. For example, `Problem checking hidden status: File has an empty path`.

Shizcow replied:
You could also return `Result<bool,std::error::Error>` so that you can mix custom string errors with io errors. `std::error::Error` implements display so you can just print and exit at top level.

benjaminedwardwebb replied:
Finally got a version that folds the `Result` type throughout. :)

Shizcow commented on src/stest/src/lib.rs:
```diff
@@ -0,0 +1,128 @@
+    /// We combine the result of each individual test with logical AND and return the result,
+    /// inverting it if necessary.
+    fn test(&self, file: &File) -> bool {
+        let passes_all_tests =
+            (!self.config.requires_each_file_is_hidden || file.is_hidden()) &&
+            (!self.config.requires_each_file_is_block_special || file.is_block_special()) &&
+            (!self.config.requires_each_file_is_character_special || file.is_character_special()) &&
+            (!self.config.requires_each_file_is_directory || file.is_directory()) &&
+            (!self.config.requires_each_file_exists || file.exists()) &&
+            (!self.config.requires_each_file_is_file || file.is_file()) &&
+            (!self.config.requires_each_file_has_set_group_id || file.has_set_group_id()) &&
+            (!self.config.requires_each_file_is_symbolic_link || file.is_symbolic_link()) &&
+            (self.optionally_test_if_newer_than_oldest_file(file)) &&
+            (self.optionally_test_if_older_than_newest_file(file)) &&
+            (!self.config.requires_each_file_is_pipe || file.is_pipe()) &&
+            (!self.config.requires_each_file_is_readable || file.is_readable()) &&
+            (!self.config.requires_each_file_has_size_greater_than_zero || file.has_size_greater_than_zero()) &&
+            (!self.config.requires_each_file_has_set_user_id || file.has_set_user_id()) &&
+            (!self.config.requires_each_file_is_writable || file.is_writable()) &&
+            (!self.config.requires_each_file_is_executable || file.is_executable());
```
If you make all of these return `Result`, you can just `?` on all of them and have this function also return a `Result<bool,String>`. That will short circuit properly.

benjaminedwardwebb replied:
I like this idea, and I've spent a little bit of time (not as much as I was hoping to, though) trying to push out the "evaluation" of the `Result` type into the outer shell of the app -- in either `main` or `lib::App.run`.

However, it's taking me some time to work around certain issues, such as the fact that `filter` can only accept a `bool` returning function, not a `Result<bool, _>`. I think this means I have to wrap the iteration itself in a `try` or `Result`-like expression to get the `Result` type back. But seems worth it to get a single-layer of error-handling at the top-level.

Shizcow commented on src/stest/src/file.rs:
```diff
@@ -0,0 +1,232 @@
+    /// See: https://stackoverflow.com/a/50045872/8732788
+    /// See: https://en.wikipedia.org/wiki/Setuid
+    pub fn has_set_group_id(&self) -> bool {
+        fn has_set_group_id(mode: u32) -> bool { mode & 0o2000 != 0 }
```
:+1:

Shizcow commented on src/stest/src/file.rs:
```diff
@@ -0,0 +1,232 @@
+}
+
+impl FromStr for File {
+    type Err = std::convert::Infallible;
```
Why is this `std::convert::Infallible`?

benjaminedwardwebb replied:
This seemed strange to me too, but apparently that is [the `Err` type for `FromStr` on `PathBuf`][1]. I guess you can `PathBuf.from_str(str)` always succeeds?

[1]: https://doc.rust-lang.org/src/std/path.rs.html#1705-1712

Shizcow commented on src/stest/src/main.rs:
```diff
@@ -0,0 +1,19 @@
+    let mut stdout = stdout().lock();
+    let result = App::new(config)
+        .run(&mut stdin, &mut stdout);
+    match result {
+        Ok(_) => ExitCode::SUCCESS,
+        Err(_) => ExitCode::FAILURE,
+    }
```
If you made all of the tests return `Result`, here is where you would print out the error before exiting. You'd just `Ok(true)` to success, `Ok(false)` to failure, and `Err(e)` to print out the error and return failure.

Shizcow reviewed (changes requested):
Sorry for just now getting to this. There are a few things that should be changed, but the overall content is great.

benjaminedwardwebb commented:
> Sorry for just now getting to this.

Not a problem at all. You're certainly under no obligation to maintain open-source projects if you're otherwise busy.

I appreciate the suggestions and comments you've made. I've responded to some of the simpler comments, and am working on addressing the issues you've called out regarding error-handling and the `checkdeps.sh` script. I should have time tomorrow to sort through some of it.

Have a good weekend. :)

Shizcow commented:
> I appreciate the suggestions and comments you've made. I've responded to some of the simpler comments, and am working on addressing the issues you've called out regarding error-handling and the `checkdeps.sh` script. I should have time tomorrow to sort through some of it.

Not a problem. If you can't/don't want to fix the checkdeps/other issues I've mentioned I can do that in a week or two myself. Looking forward to getting this merged :smile:

benjaminedwardwebb commented:
> > I appreciate the suggestions and comments you've made. I've responded to some of the simpler comments, and am working on addressing the issues you've called out regarding error-handling and the `checkdeps.sh` script. I should have time tomorrow to sort through some of it.
> 
> Not a problem. If you can't/don't want to fix the checkdeps/other issues I've mentioned I can do that in a week or two myself. Looking forward to getting this merged smile

I end up didn't having the time I thought I would last weekend to finish this. I won't be able to pick it up until next week some time, maybe next weekend.

Please feel free to merge anything else in the mean time, I'm happy to rebase over any changes. Alternatively, if you want to merge sooner you're welcome to merge it as-is and I can continue to work on the other pieces in a follow-on.

benjaminedwardwebb commented on makefile:
```diff
@@ -7,23 +7,16 @@ ifeq ($(XINERAMA),true)
-ifeq ($(CC),)
-	CC = cc
-endif
-
```
I didn't realize what failure you were talking about until returning to this just today, because `CC` happened to be set in my environment without my realizing and everything I ran seemed to work.

I've restored the bits of the `make` build that set `CC` appropriately so that the `config` subproject and `checkdeps.sh` script will work.

I agree it should be removed and would have liked to follow the nice plan you laid out, I just don't have the time or energy at the moment and wanted to get this PR into a merge-able state.

Will keep it in mind though if I find some time to work on this project again! That is, if you don't get to it first. :)

benjaminedwardwebb commented:
@Shizcow I think I've addressed your main concerns now, including the `checkdeps.sh`/`$CC` issue.

Anything holding this PR from being merged?
